### PR TITLE
fix(recorder): address custom context menus

### LINKED ISSUE #31740: [Bug]: The recorder breaks custom context menus
### Version

1.46.0-next

### Steps to reproduce

1. Run `npx playwright codegen`
2. Within the recorder's browser, navigate to a local file with the following HTML (or navigate to https://custom-context-menu.netlify.app/):

<details>
  <summary>Click to expand HTML</summary>

```html
<html>
  <head>
    <title>Custom Context Menu</title>
    <style>
      .grid {
        display: grid;
        grid-template-columns: repeat(2, 100px);
        grid-template-rows: repeat(2, 100px);
        gap: 10px;
      }
      .box {
        background-color: lightgray;
        border: 1px solid black;
        display: flex;
        justify-content: center;
        align-items: center;
      }
      #menu {
        display: none;
        flex-direction: column;
        position: absolute;
        background-color: white;
        border: 1px solid black;
      }
    </style>
  </head>
  <body>
    <div>Right click on a box to show the custom context menu.</div>
    <div class="grid">
      <div class="box">1</div>
      <div class="box">2</div>
      <div class="box">3</div>
      <div class="box">4</div>
    </div>
    <div id="menu">
      <button>Menu option 1</button>
      <button>Menu option 2</button>
    </div>
    <script>
      document.addEventListener("contextmenu", function (event) {
        const menu = document.getElementById("menu");
        if (event.target.classList.contains("box")) {
          event.preventDefault();
          menu.style.display = "flex";
          menu.style.left = `${event.pageX}px`;
          menu.style.top = `${event.pageY}px`;
        } else {
          menu.style.display = "none";
        }
      });
      document.addEventListener("click", function () {
        document.getElementById("menu").style.display = "none";
      });
    </script>
  </body>
</html>
```

</details>

3. Right click a grid cell.
4. Notice that the recorder did not generate code for the right click.
5. Left click a menu option.
6. Notice that the menu appears again (because Playwright executes the right click redundantly).

You may have to repeat steps 3 through 6 until you experience both issues.

### Expected behavior

The recorder generates code for a right click.

### Actual behavior

1. The recorder does not generate code for the right click.
2. The recorder changes the web app's behavior: **_It executes an extra right click._**

### Additional context

Video (it has sound):

https://github.com/microsoft/playwright/assets/8968171/9ef95cb7-6b60-4f8a-a0f2-b94a1fc4225f

### Environment

```shell
System:
  OS: macOS 14.5
  CPU: (12) arm64 Apple M2 Max
  Memory: 173.66 MB / 96.00 GB
Binaries:
  Node: 18.20.3 - ~/.nvm/versions/node/v18.20.3/bin/node
  npm: 10.8.1 - ~/.nvm/versions/node/v18.20.3/bin/npm
  bun: 1.0.4 - ~/.bun/bin/bun
IDEs:
  VSCode: 1.91.0 - /usr/local/bin/code
Languages:
  Bash: 3.2.57 - /bin/bash
```

### PR DESCRIPTION
**Pull request by MattKleinsmith**
resolves https://github.com/microsoft/playwright/issues/31740

## Overview of problem

The following two issues happen when using the recorder (`npx playwright codegen`) on a web app that uses a custom context menu:
1. The recorder **_does not generate code for the right click_**. (Timestamps: 0:00 - 0:15)
2. The recorder change's the web app's behavior: **_It executes an extra right click_**. (Timestamps: 0:29 - 0:40)

They happen at different times, and I don't know why one happens instead of the other.

Video (it has sound):

https://github.com/microsoft/playwright/assets/8968171/9ef95cb7-6b60-4f8a-a0f2-b94a1fc4225f

## Steps to reproduce

1. Run `npx playwright codegen`
2. Within the recorder's browser, navigate to a local file with the following HTML (or navigate to https://custom-context-menu.netlify.app/):

<details>
  <summary>Click to expand HTML</summary>

```html
<html>
  <head>
    <title>Custom Context Menu</title>
    <style>
      .grid {
        display: grid;
        grid-template-columns: repeat(2, 100px);
        grid-template-rows: repeat(2, 100px);
        gap: 10px;
      }
      .box {
        background-color: lightgray;
        border: 1px solid black;
        display: flex;
        justify-content: center;
        align-items: center;
      }
      #menu {
        display: none;
        flex-direction: column;
        position: absolute;
        background-color: white;
        border: 1px solid black;
      }
    </style>
  </head>
  <body>
    <div>Right click on a box to show the custom context menu.</div>
    <div class="grid">
      <div class="box">1</div>
      <div class="box">2</div>
      <div class="box">3</div>
      <div class="box">4</div>
    </div>
    <div id="menu">
      <button>Menu option 1</button>
      <button>Menu option 2</button>
    </div>
    <script>
      document.addEventListener("contextmenu", function (event) {
        const menu = document.getElementById("menu");
        if (event.target.classList.contains("box")) {
          event.preventDefault();
          menu.style.display = "flex";
          menu.style.left = `${event.pageX}px`;
          menu.style.top = `${event.pageY}px`;
        } else {
          menu.style.display = "none";
        }
      });
      document.addEventListener("click", function () {
        document.getElementById("menu").style.display = "none";
      });
    </script>
  </body>
</html>
```

</details>

3. Right click a grid cell.
4. Notice that the recorder did not generate code for the right click.
5. Left click a menu option.
6. Notice that the menu appears again (because Playwright executes the right click redundantly).

You may have to repeat steps 3 through 6 until you experience both issues.

## Overview of solution

Interpret `contextmenu` events as click events:

```js
onContextMenu(event: MouseEvent) {
  // in case the web page has a custom context menu
  this.onClick(event);
}
```

This solves both issues, achieving:
1. The recorder generates code for the right click.
2. The recorder does not cause an extra right click.

Timestamps: 1:57 - 2:10

## Concerns

1. The coordinates of Playwright's performed right click are a little off.
2. My knowledge of Playwright is limited, so I might be missing important edge cases.

### CODE REVIEW

**Review by yury-s — COMMENTED** [review]
Can you add a test for this? See cli-codegen-3.spec.ts for examples

**Comment by MattKleinsmith** [comment]
> Can you add a test for this? See cli-codegen-3.spec.ts for examples

https://github.com/microsoft/playwright/pull/31634/commits/aa1ca92f728296e2d56a38364665cfe106bbd549

Without the fix, the test hangs, because the recorder doesn't generate code for the right click. That is, this promise never resolves:

```js
recorder.waitForOutput('JavaScript', `button: 'right'`)
```

File: tests/library/inspector/cli-codegen-3.spec.ts [hunk 1]
```diff
@@ -559,6 +559,60 @@ await page.GetByLabel("Coun\\"try").ClickAsync();`);
...
+      recorder.waitForOutput('JavaScript', `button: 'right'`),
+      recorder.trustedClick({ button: 'right' }),
+    ]);
+    expect(message.text()).toBe('right-clicked');
+    expect(await page.evaluate('log')).toEqual([
+      // hover
+      'button: pointermove',
+      'button: mousemove',
+      // trusted right click
+      'button: pointerup',
```
Comment (MattKleinsmith) [inline tests/library/inspector/cli-codegen-3.spec.ts:605]: @yury-s I don't know why there is a `'pointerup'` here.

**Comment by MattKleinsmith** [comment]
@microsoft-github-policy-service agree company="QA Wolf"

File: tests/library/inspector/cli-codegen-3.spec.ts [hunk 2]
```diff
@@ -559,6 +559,60 @@ await page.GetByLabel("Coun\\"try").ClickAsync();`);
...
+
+    await recorder.hoverOverElement('button');
+    expect(await page.evaluate('log')).toEqual(['button: pointermove', 'button: mousemove']);
+
+    const [message] = await Promise.all([
+      page.waitForEvent('console', msg => msg.type() !== 'error'),
+      recorder.waitForOutput('JavaScript', `button: 'right'`),
+      recorder.trustedClick({ button: 'right' }),
+    ]);
+    expect(message.text()).toBe('right-clicked');
```
Comment (yury-s) [inline tests/library/inspector/cli-codegen-3.spec.ts:599]: When clicking manually on the page, the contextmenu listener is not invoked for me when recorder is on. It does get invoked without the recorder. Do you know what's going on? I'm on macbook pro, and doing ctrl+click on the trackpad.

File: tests/library/inspector/inspectorTest.ts [hunk 3]
```diff
@@ -186,9 +186,12 @@ class Recorder {
     await this.page.mouse.move(box.x + box.width / 2, box.y + box.height / 2);
   }
 
-  async trustedClick() {
-    await this.page.mouse.down();
-    await this.page.mouse.up();
+  async trustedClick(options?: {
+    button?: 'left' | 'right' | 'middle';
+    clickCount?: number;
```
Comment (yury-s) [inline tests/library/inspector/inspectorTest.ts:191]: let's drop clickCount as it is unused

File: tests/library/inspector/cli-codegen-3.spec.ts (hunk 2 above)
Comment (MattKleinsmith) [reply tests/library/inspector/cli-codegen-3.spec.ts:599]: Thank you for finding this issue. Here are some observations (with PR active):

CleanShot 2024-07-15 at 16 20 55

I will keep investigating.

File: tests/library/inspector/inspectorTest.ts (hunk 3 above)
Comment (MattKleinsmith) [reply tests/library/inspector/inspectorTest.ts]: https://github.com/microsoft/playwright/pull/31634/commits/3ece2f5aea8d745cd69a914d59d84117f8c49fdc

File: tests/library/inspector/cli-codegen-3.spec.ts (hunk 2 above)
Comment (MattKleinsmith) [reply tests/library/inspector/cli-codegen-3.spec.ts:599]: The following commit covers all three cases: https://github.com/microsoft/playwright/pull/31634/commits/5028769b527129a0722e5b9949f4bdf1ea2d2c09

1. Ctrl + click
2. Two finger click
3. External mouse right click

by having Playwright perform a right click after consuming `contextmenu`.

The previous commit did not cover the `ctrl + click` case because Playwright's performance of `ctrl + click` did not cause a `contextmenu` event, for some reason.

File: packages/playwright-core/src/server/injected/recorder/recorder.ts [hunk 4]
```diff
@@ -259,6 +259,25 @@ class RecordActionTool implements RecorderTool {
...
+      return;
+
+    this._performAction({
+      name: 'click',
+      selector: this._hoveredModel!.selector,
+      position: positionForEvent(event),
+      signals: [],
+      button: 'right',
+      modifiers: 0,
+      clickCount: event.detail
```
Comment (yury-s) [inline packages/playwright-core/src/server/injected/recorder/recorder.ts:277]: I believe `event.detail` is 0 for `contextmenu` event and we should set it to 0 here.

File: packages/playwright-core/src/server/injected/recorder/recorder.ts [hunk 5]
```diff
@@ -259,6 +259,25 @@ class RecordActionTool implements RecorderTool {
     });
   }
 
+  onContextMenu(event: MouseEvent) {
```
Comment (yury-s) [inline packages/playwright-core/src/server/injected/recorder/recorder.ts:262]: Please add a comment that the 'contextmenu' event is triggered by a right-click or equivalent action, and it prevents the click event from firing for that action, so we always convert 'contextmenu' into a right-click.

Comment (MattKleinsmith) [reply packages/playwright-core/src/server/injected/recorder/recorder.ts:262]: https://github.com/microsoft/playwright/pull/31634/commits/d9bc5fcfcce43db5870f907910f55f05c91399a4

File: packages/playwright-core/src/server/injected/recorder/recorder.ts (hunk 4 above)
Comment (MattKleinsmith) [reply packages/playwright-core/src/server/injected/recorder/recorder.ts]: https://github.com/microsoft/playwright/pull/31634/commits/6a8e68c656aa0f0e687effe0829ee74d4c99b41d

**Review by yury-s — APPROVED** [review]
Thanks for working on this and addressing all the comments! One last thing before merging this, let's follow the process and extract problem description into a GitHub issue and link to it from the PR (we usually ask all contributions to start with filing an issue).

We discussed this approach briefly in the team meeting today and one alternative idea was to dispatch `contextmenu` event directly in the generated code using `Locator.dispatchEvent`, since we don't know whether the context menu was triggered by two-finger-tap, ctrl+click or right-click. But such synthetic event would be `isTrusted=false` and ctrl+click sounds like a good enough solution which we can improve later if needed.

**Comment by MattKleinsmith** [comment]
@yury-s Thank you for your help. I have created an issue and linked the PR to it: https://github.com/microsoft/playwright/issues/31740
